Feat: Appreg v2 updates (#6855)
Kong/docs.konghq.com: app/konnect/dev-portal/applications/dynamic-client-registration/auth0.md

## Configure the Dev Portal

Once you have Auth0 configured, you can configure the Dev Portal to use Auth0 for dynamic client registration (DCR).

1. Sign in to {{site.konnect_short_name}}, then select {% konnect_icon dev-portal %} **Dev Portal** from the menu.

2. Navigate to **Application Auth** to access the authentication settings for your API Products.

3. Open the **DCR Providers** to view all configured DCR Providers

> 4. Select **New DCR Provider** button to create an Auth0 configuration. Provide a name for internal use within {{site.konnect_short_name}} and a display name for visibility on your Portal.

Reviewer: ```suggestion
4. Select **New DCR Provider** button to create an Auth0 configuration. Provide a name for internal use within {{site.konnect_short_name}}. The name and provider type information will not be exposed to Dev Portal developers.
```

5. Input the **Issuer URL** of your Auth0 tenant, formatted as: `https://AUTH0_TENANT_SUBDOMAIN.us.auth0.com`

   {:.note}
   > **Note:** You can find the value for your `AUTH0_TENANT_SUBDOMAIN` by checking the **Tenant Name** under **Settings** > **General** tab.

6. Choose Auth0 as the **Provider Type**.

> 7. Enter the Client ID from the admin application created in Auth0 above into the **Initial Client ID** field. Enter the Client secret from the admin application created in Auth0 above into the **Initial Client Secret** field.

Reviewer: ```suggestion
7. Enter the Client ID from the admin application created in Auth0 above into the **Initial Client ID** field. Enter the Client secret from the admin application created in Auth0 above into the **Initial Client Secret** field. Note: The Initial Client Secret will be stored in isolated, encrypted storage and will not be readable through any Konnect API.
```

## Configure the Dev Portal

Once you have Auth0 configured, you can configure the Dev Portal to use Auth0 for dynamic client registration (DCR).

{% navtabs %}
{% navtab Konnect UI %}

1. Sign in to {{site.konnect_short_name}}, then select {% konnect_icon dev-portal %} **Dev Portal** from the menu.

> 2. Navigate to **Application Auth** to access the authentication settings for your API Products.

Reviewer 2: ```suggestion
2. Navigate to **Application Auth** to access the authentication settings for your API products.
```

> 3. Open the **DCR providers** to view all configured DCR providers

Reviewer 2: ```suggestion
3. Open **DCR providers** to view all configured DCR providers.
```

> 4. Select **New DCR provider** to create an Auth0 configuration. Provide a name for internal use within {{site.konnect_short_name}}. The name and provider type information will not be exposed to Dev Portal developers.

Reviewer 2: ```suggestion
4. Select **New DCR provider** to create an Auth0 configuration. Provide a name for internal use in {{site.konnect_short_name}}. The name and provider type information will not be exposed to Dev Portal developers.
```

   {:.note}
   > **Note:** You can find the value for your `AUTH0_TENANT_SUBDOMAIN` by checking the **Tenant Name** under **Settings** > **General**.

6. Choose Auth0 as the **Provider Type**.

7. Input the Client ID of the previously created admin application in Auth0 into the **Initial Client ID** field. input the Client Secret of the same Auth0 admin application into the Initial Client Secret field. **The Initial client secret will be stored in isolated, encrypted storage and will not be readable through any Konnect API.**

8. Enter the **Client Token Audience** as the identifier value you set when configuring the API entity above. If you’re using developer managed scopes, this value should map to the audience field of your associated Auth0 API.

9. **Optional**: If you are using developer-managed scopes, select the **Use Developer Managed Scopes** checkbox.

10. Save the DCR provider. You should now see it in the list of DCR providers.

11. Click the **Auth Strategy** tab to see all of the auth strategies. Select **New Auth Strategy** to create an auth strategy that uses the DCR provider you created.

> 12. Enter a name for internal use in {{site.konnect_short_name}} and a display name that will be displayed the portal. In the **Auth Type** dropdown menu select DCR. In the **DCR provider** dropdown, select the name of the DCR provider config you created. Your **Issuer URL** will be prepopulated with the Issuer URL you added to the DCR provider.

Reviewer 2: ```suggestion
12. Enter a name for internal use in {{site.konnect_short_name}} and a display name that will be displayed in Dev Portal. In the **Auth Type** dropdown menu select DCR. In the **DCR provider** dropdown, select the name of the DCR provider config you created. Your **Issuer URL** will be prepopulated with the Issuer URL you added to the DCR provider.
```

> 13. Enter the mandatory `openid` scope into the **Scopes** field in addition to any other scopes your developers may need access to (e.g. `openid, read:account_information, write:account_information`, etc). If you’re using developer-managed scopes, these are the scopes your developers will be able to _choose_ from in Dev Portal.

Reviewer 2: ```suggestion
13. Enter the mandatory `openid` scope into the **Scopes** field in addition to any other scopes your developers may need access to (for example, `openid, read:account_information, write:account_information`). If you’re using developer-managed scopes, these are the scopes your developers will be able to _choose_ from in Dev Portal.
```

14. Enter `azp` into the **Credential Claims** field, which will match the client ID of each Auth0 application

> 15. Choose the **Auth Methods** required (`client_credentials`, bearer, session), and **save**.

Reviewer 2: ```suggestion
15. Choose the required **Auth Methods** (`client_credentials`, bearer, session) and **save**.
```

{% endnavtab %}
{% navtab API %} 
After configuring Auth0, you can integrate it with the Dev Portal for dynamic client registration (DCR). This process involves two steps: creating the DCR provider and establishing the authentication strategy. DCR providers are designed to be reusable configurations. This means once you've configured the Auth0 DCR provider, it can be utilized across multiple authentication strategies without needing to be set up again.

> 1. Start by creating the DCR provider. Send a `POST` request to the[`v2/dcr-providers`](https://kong-platform-api.netlify.app/konnect/application-auth-strategies/v2/openapi.yaml/#tag/DCR-Providers/operation/create-dcr-provider) endpoint with your DCR configuration details. 

Reviewer 2: ```suggestion
1. Start by creating the DCR provider. Send a `POST` request to the [`v2/dcr-providers`](https://kong-platform-api.netlify.app/konnect/application-auth-strategies/v2/openapi.yaml/#tag/DCR-Providers/operation/create-dcr-provider) endpoint with your DCR configuration details:
```

```sh
   curl --request POST \
   --url https://us.api.konghq.com/v2/dcr-providers \
   --header 'Authorization: $KPAT' \
   --header 'content-type: application/json' \
   --data '{
   "name": "Auth0 DCR Provider",
   "provider_type": "auth0",
   "issuer": "https://my-issuer.auth0.com/api/v2/",
   "dcr_config": {
      "initial_client_id": "abc123",
      "initial_client_secret": "abc123xyz098!",
      "initial_client_audience": "https://my-custom-domain.com/api/v2/"

>    }

Reviewer 2: I think the spacing might be off on this one

   }'
```
You will receive a response that includes a `dcr_provider` object similar to the following:

   ```sh
   {
      "created_at": "2024-02-29T23:38:00.861Z",
      "updated_at": "2024-02-29T23:38:00.861Z",
      "id": "93f8380e-7798-4566-99e3-2edf2b57d289",
      "name": "Auth0 DCR Provider",
      "provider_type": "auth0",
      "issuer": "https://my-issuer.auth0.com/api/v2/",
      "dcr_config": {
         "initial_client_id": "abc123",
         "initial_client_audience": "https://my-custom-domain.com/api/v2/"
      },
      "active": false
   }

   ```
Save the `id` value for creating the authentication strategy.

> 2. With the `dcr_id` obtained from the first step, create an authentication strategy. Construct a `POST` request to the [`v2/create-auth-stratgies`](https://kong-platform-api.netlify.app/konnect/application-auth-strategies/v2/openapi.yaml/#tag/App-Auth-Strategies/operation/create-app-auth-strategy) endpoint describing an authentiaction strategy: 

Reviewer 2: ```suggestion
2. With the `dcr_id` obtained from the first step, create an authentication strategy. Construct a `POST` request to the [`v2/create-auth-stratgies`](https://kong-platform-api.netlify.app/konnect/application-auth-strategies/v2/openapi.yaml/#tag/App-Auth-Strategies/operation/create-app-auth-strategy) endpoint describing an authentication strategy: 
```
